core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -743,19 +743,23 @@ private PartitionSpec reassignPartitionIds(PartitionSpec partitionSpec, TypeUtil
 
     } else {
       // for v1, preserve the existing spec and carry forward all fields, replacing missing fields with void
-      Map<Pair<Integer, String>, PartitionField> newFields = Maps.newLinkedHashMap();
+      Map<Integer, PartitionField> newFields = Maps.newLinkedHashMap();
       for (PartitionField newField : partitionSpec.fields()) {
-        newFields.put(Pair.of(newField.sourceId(), newField.transform().toString()), newField);
+        newFields.put(newField.sourceId(), newField);
       }
 
       for (PartitionField field : spec().fields()) {
         // ensure each field is either carried forward or replaced with void
-        PartitionField newField = newFields.remove(Pair.of(field.sourceId(), field.transform().toString()));
```

Comment:
Why did you remove the transform string from the dedup? The fields are different if they have different transforms so I don't think that this is correct. For example, if the current spec is `[ 1000: bucket[16](2: uuid) ]` and the incoming spec is `[ 1000: identity(2: uuid) ]` then the result must be `[ 1000: always_null(2: uuid), 1001: identity(2: uuid) ]`. That's because the output of `identity(uuid)` is a string and the output of `bucket[16](uuid)` is an int. If we tried to read the files table after this change, it would fail.

Reply:
Sorry this is from a bad merge, did push to oss right before I gave another talk on friday. This is part of me originally tinkering around.

Reply:
Huh some how I pushed but github did not realize it was this branch